trieste/models/keras/architectures.py

```diff
@@ -0,0 +1,315 @@
+        self._model = self.build_ensemble()
+
+    def __repr__(self) -> str:
+        """"""
+        return f"KerasEnsemble({self._networks!r})"
+
+    @property
+    def model(self) -> tf.keras.Model:
+        """ " Returns built but uncompiled Keras ensemble model."""
+        return self._model
+
+    @property
+    def ensemble_size(self) -> int:
+        """
+        Returns the size of the ensemble, that is, the number of base learners or individual neural
+        network models in the ensemble.
+        """
+        return len(self._networks)
+
+    def build_ensemble(self) -> tf.keras.Model:
+        """
+        Builds the ensemble model by combining all the individual networks in a single Keras model.
+        This method relies on ``connect_layers`` method of :class:`KerasEnsembleNetwork` objects
+        to construct individual networks.
+
+        :return: The Keras model.
+        """
+        inputs = []
+        outputs = []
+        for index, network in enumerate(self._networks):
+            network.network_name = "model_" + str(index) + "_"
+            input_tensor, output_tensor = network.connect_layers()
+            inputs.append(input_tensor)
+            outputs.append(output_tensor)
+
+        return tf.keras.Model(inputs=inputs, outputs=outputs)
+
+
+class KerasEnsembleNetwork(tf.Module):
+    """
+    This class is an interface that defines necessary attributes and methods for neural networks
+    that are meant to be used for building ensembles by
+    :class:`~trieste.models.keras_networks.KerasEnsemble`. Subclasses are not meant to
+    build and compile Keras models, instead they are providing specification that
+    :class:`~trieste.models.keras_networks.KerasEnsemble` will use to build the Keras model.
+    """
+
+    def __init__(
+        self,
+        input_tensor_spec: tf.TensorSpec,
+        output_tensor_spec: tf.TensorSpec,
+        network_name: str = "",
+    ):
+        """
+        :param input_tensor_spec: Tensor specification for the input to the network.
+        :param output_tensor_spec: Tensor specification for the output of the network.
+        :param network_name: The name to be used when building the network.
+        """
+        if not isinstance(input_tensor_spec, tf.TensorSpec):
+            raise ValueError(
+                f"input_tensor_spec must be an instance of tf.TensorSpec, "
+                f"received {type(input_tensor_spec)} instead."
+            )
+        if not isinstance(output_tensor_spec, tf.TensorSpec):
+            raise ValueError(
+                f"output_tensor_spec must be an instance of tf.TensorSpec, "
+                f"received {type(output_tensor_spec)} instead."
+            )
+
+        self.input_tensor_spec = input_tensor_spec
+        self.output_tensor_spec = output_tensor_spec
+        self.network_name = network_name
+
+    @property
+    def input_layer_name(self) -> str:
+        return self.network_name + "input"
```

Comment:
(wouldn't it look nicer to add a separator?)

```suggestion
        return self.network_name + "_input"
```

Reply:
thats done when changing the `network_name`